### PR TITLE
finished back-porting parser-improvement changes, need to check win64…

### PR DESCRIPTION
**Pull request by jonathan-beard**
In progress, will set to draft momentarily. 

## In this pull request
* changes for the fifo, ringbuffers, blocked structures
* aim is to make all structures with the ringbuffer compatible 
once again with shared memory so we can go back to multiprocess
in addition to multi-thread. 
* changes should also make it amenable to make the fifo itself a module, and 
therefore be able to insert whatever FIFOs we want to (e.g. #zeromq, https://github.com/zeromq/libzmq) and boostfifo, even MPI with wrappers. 

@f-tischler, could you look over chances with an eye towards win, I'm gonna double check momentarily to make doubly sure I didn't miss any Win specific guards that were supposed to be in place.

### CODE REVIEW

**Comment by jonathan-beard** [comment]
bufferdata.tcc  is definitely one guard I missed, include file...shouldn't be included, will fix in a moment.

**Comment by f-tischler** [comment]
Ill take a look at it tomorrow. Hopefully I will also get to finish work for my other PR regading cache line size detection on Win

**Comment by jonathan-beard** [comment]
> Ill take a look at it tomorrow. Hopefully I will also get to finish work for my other PR regading cache line size detection on Win

I already found a few things I missed when back-porting. Will fix tonight, hopefully before you hop on to take a look. Technically bank-holiday today, so...plenty of time to work on fun side projects for me :)

File: raftinc/sysschedutil.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * yield - generic yield function for whatever the underlying
+ * implementation is, could be qthreads, a process, or a thread
+ * but it'll call the right implementation for you. 
+ */
+static inline void yield()
+{
+
+#if (! defined _WIN64) || (! defined _WIN32)
```
Comment (jonathan-beard) [inline raftinc/sysschedutil.hpp:42]: Needs to be an and not or

**Comment by jonathan-beard** [comment]
Ok, @f-tischler, I think this one is now finally actually ready for review.

**Comment by f-tischler** [comment]
Seems fine to me and since it's building and passing tests I'd say you are good to go @jonathan-beard

File: raftinc/database.tcc [hunk 2]
```diff
@@ -40,86 +41,68 @@ namespace Buffer
...
-   /** variable set by scheduler, used for shutdown **/
-   bool                    is_valid       = true;
+    /**
+     * read/write pointer. the thread_access is in 
+     * between as it's well accessed by both just
+     * as frequently as the pointers themselves, 
+     * so we get decent caching behavior out of 
+     * doing it this way. 
+     */
+    Pointer                 read_pt;
```
Comment (jonathan-beard) [inline raftinc/database.tcc:77]: Need to fix alignment on these.....each should be on a unique cache line

File: raftinc/database.tcc [hunk 3]
```diff
@@ -40,86 +41,68 @@ namespace Buffer
...
+    ThreadAccess            thread_access[ 2 ];
+    Pointer                 write_pt;
+    
+    /**
```
Comment (jonathan-beard) [inline raftinc/database.tcc:81]: Delete comment

File: raftinc/database.tcc [hunk 4]
```diff
@@ -40,86 +41,68 @@ namespace Buffer
...
+    bool                    is_valid        = true;
+    
+    /**
+     * these keep reference over how many read/writes are
+     * blocked. used for dynamic adaptation.
+     */
+    Blocked                 read_stats; 
+    Blocked                 write_stats;
+    
+    /** need to force resize, this has the count requested **/
```
Comment (jonathan-beard) [inline raftinc/database.tcc:98]: Needs test case for forced resize

File: raftinc/datamanager.tcc [hunk 5]
```diff
@@ -223,25 +229,20 @@ private:
 
    bool                  resizeable          = true;
    
-   struct ThreadAccess
-   {
-      union
-      {
-         std::uint64_t whole = 0; /** just in case, default zero **/
-         dm::key_t     flag[ 8 ];
-      };
```
Comment (jonathan-beard) [inline raftinc/datamanager.tcc:232]: Replace padding with template padding solution

File: raftinc/defs.hpp [hunk 6]
```diff
@@ -1,9 +1,9 @@
 /**
  * defs.hpp - 
  * @author: Jonathan Beard
- * @version: Mon Sep  3 16:20:48 2018
+ * @version: Sun Feb  7 05:46:48 2016
```
Comment (jonathan-beard) [inline raftinc/defs.hpp:4]: Update copyright, need to change all to raftlib org

File: raftinc/fifoabstract.tcc [hunk 7]
```diff
@@ -25,19 +25,62 @@
...
+#include "internaldefs.hpp"
 
 template < class T, Type::RingBufferType type > 
    class FIFOAbstract : public FIFO
 {
 public:
-   FIFOAbstract() : FIFO()
-   {
-   }
+   FIFOAbstract() : FIFO(){}
```
Comment (jonathan-beard) [inline raftinc/fifoabstract.tcc:36]: Consider using default syntax

File: raftinc/pointer.hpp [hunk 8]
```diff
@@ -1,9 +1,9 @@
 /**
  * pointer.hpp - 
  * @author: Jonathan Beard
- * @version: Thu May 15 09:58:51 2014
+ * @version: Sun Oct 30 04:58 2016
  * 
- * Copyright 2014 Jonathan Beard
```
Comment (jonathan-beard) [inline raftinc/pointer.hpp:6]: Copyright needs fixing

File: raftinc/pointer.hpp [hunk 9]
```diff
@@ -25,17 +25,21 @@
...
 public:
+   Pointer() : max_cap( 0 ){}
+
    /**
     * Pointer - used to synchronize read and write
     * pointers for the ring buffer.  This class encapsulates
     * wrapping.
     */
-   Pointer( const std::size_t cap ) : max_cap( cap ){};
+   Pointer( const std::size_t cap ) : max_cap( cap ){}
```
Comment (jonathan-beard) [inline raftinc/pointer.hpp:42]: Make constexpr if possible

File: raftinc/pointer.hpp [hunk 10]
```diff
@@ -46,21 +50,21 @@ class Pointer{
     * @param   other, const Pointer&, the other pointer to be cpied
     * @param   new_cap, the new max cap
     */
-   Pointer( Pointer * const other, 
+   Pointer( Pointer &other,
```
Comment (jonathan-beard) [inline raftinc/pointer.hpp:53]: Make param const

File: raftinc/ringbuffer.tcc [hunk 11]
```diff
@@ -57,7 +61,9 @@ public:
         : RingBufferBase<T, type>()
     {
         assert(n != 0);
-        (this)->datamanager.set(new Buffer::Data<T, type>(n, align));
```
Comment (jonathan-beard) [inline raftinc/ringbuffer.tcc:60]: Fixing per align will fix this

File: raftinc/ringbufferbase.tcc [hunk 12]
```diff
@@ -56,14 +56,15 @@ template < class T,
 class RingBufferBase
 {
 public:
-    RingBufferBase() = delete;
-    virtual ~RingBufferBase() = delete;
+    RingBufferBase() = default;
```
Comment (jonathan-beard) [inline raftinc/ringbufferbase.tcc:59]: Why delete vs default?

File: raftinc/sysschedutil.hpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * yield - generic yield function for whatever the underlying
+ * implementation is, could be qthreads, a process, or a thread
+ * but it'll call the right implementation for you. 
+ */
+static inline void yield()
+{
+
+#if (! defined _WIN64) && (! defined _WIN32)
```
Comment (jonathan-beard) [inline raftinc/sysschedutil.hpp:42]: Need to figure out windows actions here

**Pull request merged**